Readability.js

```diff
@@ -867,6 +878,41 @@ Readability.prototype = {
           lastScore = parentOfTopCandidate.readability.contentScore;
           parentOfTopCandidate = parentOfTopCandidate.parentNode;
         }
+
+        // Find a better top candidate node if it contains (at least three) nodes which belong to `topCandidates` array
+        // and whose scores are quite closed with current `topCandidate` node.
+        var topCandidateAncestors = [];
+        topCandidates.some(function(_topCandidate, i) {
+          if (i === 0) {
+            return false;
+          }
+          if (Math.abs((_topCandidate.readability.contentScore -
+                        topCandidate.readability.contentScore) /
+                        topCandidate.readability.contentScore) < 0.25) {
+            topCandidateAncestors.push(this._getNodeAncestors(_topCandidate));
+          }
+          return false;
+        }.bind(this));
+
+        topCandidateAncestors.some(function(ancestors, i) {
+          var that = this;
+          return that._someNode(ancestors, function(topCandidateAncestor) {
+            var numberOfHighScoreNode = 0;
+            return topCandidateAncestors.some(function(_ancestors, j) {
```

Comment:
You're using 3 nested loops and another `indexOf` check. You shouldn't need that many loops. Instead, first check that we found at least 3 nodes, then loop over the ancestors of the top candidate and check if the list you built up of the ancestors of the other candidates has at least 3 sub-lists that contain the ancestor:

```js
var MINIMUM_TOPCANDIDATES = 3;
if (alternativeCandidateAncestors.length >= MINIMUM_TOPCANDIDATES) {
  var topCandidateAncestor = topCandidate.parentNode;
  while (topCandidateAncestor) {
    var listsContainingThisAncestor = 0;
    for (var ancestorIndex = 0; ancestorIndex < alternativeCandidateAncestors.length && listsContainingThisAncestor < MINIMUM_TOPCANDIDATES; ancestorIndex++) {
       listsContainingThisAncestor += Number(ancestorsForOneNode.includes(topCandidateAncestor));
    }
    if (listsContainingThisAncestor == MINIMUM_TOPCANDIDATES) {
      topCandidate = topCandidateAncestor;
      break;
    }
    topCandidateAncestor = topCandidateAncestor.parentNode;
  }
}
```